src/packages/@ncigdc/utils/auth/awg.js

```diff
@@ -0,0 +1,137 @@
+import { throttle } from 'lodash';
+import urlJoin from 'url-join';
+import moment from 'moment';
+
+import consoleDebug from '@ncigdc/utils/consoleDebug';
+import {
+  AWG,
+  AWG_TOKEN_EXPIRY, // defaults to 5 minutes
+  AWG_TOKEN_RENEWAL_INTERVAL, // defaults to 0, disabled
+  FENCE,
+} from '@ncigdc/utils/constants';
+
+const createSessionRenewalTimestamp = now => {
+  consoleDebug(`Creating session timestamp at ${now}`);
+  localStorage.setItem('awg_sessionLastRenewed', now);
+  return !!localStorage.awg_sessionLastRenewed;
+};
+
+const removeSessionRenewalTimestamp = () => {
+  localStorage.removeItem('awg_sessionLastRenewed');
+  return !localStorage.awg_sessionLastRenewed;
+};
+
+const resetSessionRenewalTimestamp = now => (
+  removeSessionRenewalTimestamp() &&
+  createSessionRenewalTimestamp(now)
+);
+
+/* this is meant to avoid duplicate queries, because having multiple Relay Environments
+ * means renewal race conditions not worth dealing with at this point. ¯\_(ツ)_/¯
+ */
+let queryCache = null;
+const handleQueryCache = value => {
+  // hold and use the latest response for 5 seconds;
+  queryCache = value;
+
+  setTimeout(() => {
+    queryCache = null;
+  }, 5000);
+
+  return queryCache;
+};
+
+const renewAWGToken = throttle(({
+  now,
+  origin,
+  timeSinceLastRenewal,
+}) => {
+  consoleDebug(`Querying AWG fence at ${now}, by ${origin}`);
+
+  return queryCache || fetch(urlJoin(FENCE, 'user'), {
```

Comment:
`queryCache` could be `err` from the line 74, which will be truthy.  But `queryCache` could be falsey from line 63 when the fence request fails.  Is it intentional this to behave like a five-second debounce for thrown errors but not debounce when the previous request is not `ok`?

Reply:
That's right. The idea here (if you don't mind me breaking it down, just to double check my logic) is that if `queryCache` is `null`, it means a request hasn't happened in the last 5 seconds. 
If instead it has "any" value, then there's another request "going on" (not really, as I'm straight up giving the response if available...), so don't make a new call.
In other words, I'm micromanaging asynchronicity and runtime delays created by technical debt.
Part of the reason why I chose to update some of the dependencies.

How it happens:
There cases in which multiple UI components are making almost simultaneous requests, with a second or less of difference in between them, which are not getting caught by `lodash.throttle`, before the first one updates the timestamp to prevent the other requests from happening. If I extend the throttling, then those other attempts are rejected, and the whole app flow crashes, logging the user out immediately, even though the fence validates the first call.
This is as close as a sweet spot as I managed to find.

To your question, the result itself being `truthy` or `falsy` or even an error, is not a concern here, as the respective handling is happening elsewhere. It's more avoiding the duplication without collapsing the chain of actions dispatched.

Does any of this make sense? lol

Reply:
The explanation makes sense to me. It seems like `_.debounce` should cover this, but it sounds like Lodash wasn't reliably grouping the calls for some reason. I'll think about it a little more and see if I'm maybe missing something, though.

Reply:
No, you're right. Somehow lodash is not grouping correctly all the calls made from different callers.
My best guess is it is some sort of transpilation issue, in which the throttled function is getting different identifiers in different contexts, and Lodash is not associating them as the same.

Could not test this theory locally (pre-transpilation) since we are not able/allowed to run auth on local machine.